feat: updating to Style Dictionary v4

PKulkoRaccoonGang opened the pull request:
## Description

This PR is an attempt to update the Style Dictionary v3 library to Style Dictionary v4 (including refactoring the design tokens to the DTCG format) + fixed some problems with typography.

## Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [ ] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please request an a11y review for the PR in the [#wg-paragon](https://openedx.slack.com/archives/C02NR285KV4) Open edX Slack channel.
* [ ] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [ ] Is there adequate test coverage for your changes?

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

PKulkoRaccoonGang committed:
feat: update style dictionary to v4

PKulkoRaccoonGang committed:
feat: converted utils and breakpoints

PKulkoRaccoonGang committed:
feat: DTCG format

openedx-webhooks commented:
Thanks for the pull request, @PKulkoRaccoonGang!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/paragon-working-group`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

PKulkoRaccoonGang commented:
**TODO**
- correct display of the header of generated files
- checking the correctness of variable generation (for some variables you can notice an extra indication of the unit of measurement, for example 0rem)
 ...

**ES6 modules**
Since the updated Style Dictionary v4 API is based on [ES6 modules](https://v4.styledictionary.com/version-4/migration/#file-headers), an open question is finding a solution for the correct operation of Style Dictionary v4 together with the existing CommonJS functionality of Gatsby. According to [the Gatsby documentation](https://www.gatsbyjs.com/docs/how-to/custom-configuration/es-modules/), support for ES6 modules is implemented in more recent releases of the framework. Most likely, for the current version of Gatsby (`4.23.1`), a potential solution could be to use asynchronous imports or update the Gatsby version.

brian-smith-tcril commented:
Thank you so much for putting this together! Considering the task of "do some discovery around v4" this is truly above and beyond!

> checking the correctness of variable generation (for some variables you can notice an extra indication of the unit of measurement, for example 0rem)

I think it's reasonable to update the `.json` files to include units where we don't yet have them as a solution here.

> Since the updated Style Dictionary v4 API is based on [ES6 modules](https://v4.styledictionary.com/version-4/migration/#file-headers), an open question is finding a solution for the correct operation of Style Dictionary v4 together with the existing CommonJS functionality of Gatsby. According to [the Gatsby documentation](https://www.gatsbyjs.com/docs/how-to/custom-configuration/es-modules/), support for ES6 modules is implemented in more recent releases of the framework. Most likely, for the current version of Gatsby (4.23.1), a potential solution could be to use asynchronous imports or update the Gatsby version.

Could you elaborate on this a bit? I worry that moving from Gatsby v4 to v5 would be quite a heavy lift, and I wonder how that compares to some workarounds people have tried https://github.com/gatsbyjs/gatsby/discussions/31599#discussioncomment-1239988

PKulkoRaccoonGang committed:
refactor: added dynamic imports

PKulkoRaccoonGang committed:
refactor: code refactoring

PKulkoRaccoonGang committed:
refactor: corrected header

adamstankiewicz commented on package.json:
```diff
@@ -198,10 +198,10 @@
   "lint-staged": {
-    "*.{js,jsx,ts,tsx}": [
+    "*.{mjs}": [
       "npx eslint"
     ],
-    "*.scss": [
+    "*.sass": [
```
[curious] Can you clarify these changes a bit? Why are they needed?

PKulkoRaccoonGang replied:
I needed to ignore the pre-commit hook)
This is a temporary change that I will remove 💯

adamstankiewicz commented on styles/css/core/variables.css:
```diff
@@ -70,26 +78,25 @@
   --pgn-typography-font-size-lg: 1.4063rem;
   --pgn-typography-font-size-base: 1.125rem;
-  --pgn-typography-font-family-monospace: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono",  "Courier New", monospace;
+  --pgn-typography-font-family-monospace: SFMono-Regular, Menlo, Monaco, Consolas, '"Liberation Mono"', '"Courier New"', monospace;
```
[clarification] Is having the already quoted font names now wrapped with single quotes, too intended/needed/work as expected?

PKulkoRaccoonGang replied:
Resolved in https://github.com/openedx/paragon/pull/3207

adamstankiewicz commented on tokens/src/core/components/ActionRow.json:
```diff
@@ -1,9 +1,10 @@
 {
+  "$type": "dimension",
```
[clarification] Will _all_ tokens defined for `ActionRow` be of `"$type": "dimension"`? Should this be associated more directly to the dimension-specific tokens, even though these are the only tokens in this file right now?

For example, if/when a non-dimension token is added to this file, is there a risk this `$type` won't be semantically correct/accurate anymore?

Similar question like applies to other components' tokens, too.

PKulkoRaccoonGang replied:
Resolved in https://github.com/openedx/paragon/pull/3207

adamstankiewicz commented on tokens/src/core/components/Card.json:
```diff
@@ -1,95 +1,82 @@
-          }
+          "base": { "source": "$card-border-radius", "$value": "{size.border.radius.base}" },
+          "logo": { "source": "$card-logo-border-radius", "$value": ".25rem" },
+          "inner": { "source": "$card-inner-border-radius", "$value": "calc({size.card.border.radius.base} - {size.card.border.width})" }
```
[curious] Have we looked into whether we can avoid including CSS-specific syntax `calc` in the token itself, or how such tokens would be transformed outside of CSS variables?

For example, if there is arithmetic like subtraction as in this line, could the token `$value` be `{size.card.border.radius.base} - {size.card.border.width}` and have some sort of transform for the CSS variables output to wrap it with `calc(...)`?

That way, tokens like these would be more portable to other non-CSS based platforms like mobile.

adamstankiewicz replied:
Dug into this a bit. It appears this still remains a complex problem within the `style-dictionary` community. There is support for a `resolveMath` transform, from `tokens-studio/sd-transforms`; however, when the token outputs references via `outputReferences: true | fn`, the transformed/resolved math is replaced with the original reference (i.e., a CSS variable with `var(--foo: 5px)` in this case).

I attempted to craft a solution for https://github.com/tokens-studio/sd-transforms/issues/203, but came to the same realization recently described [here](https://github.com/amzn/style-dictionary/issues/1055#issuecomment-2312072149):

> The big issue is that outputting refs happens on the format lifecycle and wrapping values in calc() statements usually happens in the transform lifecycle that happens before. So even if you wrap the values in calc() in transform, the format part will just undo that work by outputting refs by using the original value.
>
> Either the calc wrapping needs to happen on the format level or the outputting refs util needs to act on the transformed value somehow (keeping in mind the bugs that this used to cause in v3 and not regressing on this again)
>
> There are a couple of open issues to rethink how references are resolved and how values with references in them are transformed for a future v5 version of Style Dictionary, but this topic is rather complex and this issue won't be fixed until we have solutions to that broader topic.

tl;dr; Handling the wrapping of math expressions with the CSS `calc` syntax remains an open question, and likely will not land in `v4` of Style Dictionary per the above. Given this, it probably makes sense to keep the `calc` syntax for now, but when we begin transforming to non-CSS platforms (e.g., JavaScript, iOS, Android), having `calc` in the token value itself will present an issue. A possible workaround for if/when this becomes an issue might be to introduce custom transforms for the (future) non-CSS platforms that strips the wrapping `calc(...)` from the underlying math operations.

PKulkoRaccoonGang replied:
This is definitely a very interesting question that we haven't thought about yet. Thanks for your research, now we know more!

adamstankiewicz commented on tokens/src/core/components/general/link.json:
```diff
@@ -1,27 +1,28 @@
 {
+  "$type": "textDecoration",
```
[clarification] Similar to other comment about the `dimension` type defined more globally in a file, will _all_ tokens in this file be related to text-decoration? It would seem perhaps the `$type` should be defined within the `decoration` sub-object?

PKulkoRaccoonGang replied:
Resolved in https://github.com/openedx/paragon/pull/3207

adamstankiewicz commented on tokens/src/core/global/other.json:
```diff
@@ -1,4 +1,8 @@
 {
+  "$type": "percentage",
```
[clarification] only the `theme-interval` token specifically in this file is a percentage, right?

PKulkoRaccoonGang replied:
Resolved in https://github.com/openedx/paragon/pull/3207

adamstankiewicz commented on tokens/src/core/global/typography.json:
```diff
@@ -2,89 +2,222 @@
+        "base": {
+          "source": "$font-size-base",
+          "$value": "1.125rem",
+          "$type": "dimension",
```
[suggestion] could the `"$type": "dimension"`, be moved up under `size` so it applies to all size permutations (e.g., base vs. lg vs. sm)?

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3203

adamstankiewicz commented on tokens/src/themes/light/components/Avatar.json:
```diff
@@ -1,7 +1,8 @@
 {
+  "$type": "color",
   "color": {
```
[curious] Based on the [migration guidelines](https://styledictionary.com/version-4/migration/#cti-reliance):

<img width="917" alt="image" src="https://github.com/user-attachments/assets/eec5cdba-4899-4a27-9787-433cdc8a7faa">

It _seems_ the top-level `"color"` key here isn't actually needed anymore for `style-dictionary` to determine the CTI (Category-Type-Item), given the `$type` annotation. Is this something that was explored? It might help simplify JSON files slightly by removing possible layer(s) of nesting. If it is possible to nix the extra nesting layer, curious to hear your thoughts around scope in tackling that broadly.

adamstankiewicz replied:
Note: I might be misinterpreting this, where the top-level `color` is still needed from a grouping perspective; just not from a type pov.

adamstankiewicz commented on tokens/style-dictionary.js:
```diff
@@ -65,165 +80,188 @@ const colorTransform = (token, theme) => {
   themeVariant,
-}) => {
+}, StyleDictionary) => {
+  // eslint-disable-next-line import/no-unresolved
+  const { sortByReference, usesReferences, getReferences } = (await import('style-dictionary/utils'));
```
I wonder if it'd make sense to abstract a helper util function to make these utils accessible, e.g. 

```jsx
/* eslint-disable import/no-unresolved */
const getStyleDictionaryUtils = async () => import('style-dictionary/utils');
/* eslint-enabled import/no-unresolved */

// createCustomCSSVariables
const { sortByReference, usesReferences, getReferences } = await getStyleDictionaryUtils();
```

`getStyleDictionaryUtils` could also be used within `initializeStyleDictionary`, e.g.:

```jsx
const initializeStyleDictionary = async () => {
  // eslint-disable-next-line import/no-unresolved
  const StyleDictionary = (await import('style-dictionary')).default;
  const { getReferences } = await getStyleDictionaryUtils();
 
  // ...
}
```

adamstankiewicz replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3203

adamstankiewicz commented on tokens/src/core/global/spacing.json:
```diff
@@ -1,36 +1,89 @@
+        "$value": "1rem",
+        "$description": "Basic space value"
+      },
+      "1,5": {
```
[clarification] Is the comma in `"1,5"` and other spacing tokens intended, or should it be a period (i.e., `"1.5"`)?

PKulkoRaccoonGang replied:
It doesn't matter for the final result. But I agree with you, I replaced the comma with a period (by analogy to the value in the description)

[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3207

adamstankiewicz commented on tokens/style-dictionary.js:
```diff
@@ -65,165 +80,188 @@ const colorTransform = (token, theme) => {
     ? dictionary.allTokens.filter(token => token.filePath.includes(themeVariant))
     : dictionary.allTokens;
 
   const variables = outputTokens.sort(sortByReference(dictionary)).map(token => {
```
Related, upon further investigation, it looks we _do_ rely on token-specific handling of `outputReferences`. For example, we opt-out of including references even when the formatter has them opted-in (e.g., for the case of always including HEX values in SVG `url` values).

That said, token-specific `outputReferences` is still supported with the out-of-the-box `formattedVariables`, by passing `outputReferences` a function instead of a boolean, e.g.:

```jsx
const variables = formattedVariables({
  format: 'css',
  dictionary,
  outputReferences: (token) => {
    // Formatter options suggest to never output references
    if (!outputReferences) {
      return false;
    }
    // Formatter options suggest to output references, but individual token might opt-out
    if (!token.outputReferences) {
      return false;
    }
    // Otherwise, output references.
    return true;
  },
  usesDtcg: true,
});
```

adamstankiewicz replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3203

adamstankiewicz commented on tokens/style-dictionary.js:
```diff
@@ -65,165 +80,188 @@ const colorTransform = (token, theme) => {
+        '/*',
+        ' * IMPORTANT: This file is the result of assembling design tokens.',
+        ' * Do not edit directly.',
+        ` * Generated on ${currentDate}`,
```
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3203

PKulkoRaccoonGang committed:
chore: corrected design tokens types

PKulkoRaccoonGang committed:
chore: the fractional value of the spacer has had its comma replaced with a period

PKulkoRaccoonGang committed:
refactor: corrected changed values of tokens and CSS variables

adamstankiewicz committed:
feat: --output-references CLI arg, registers filters, updates CSS vars format (#3203)

* feat: --output-references CLI arg for build-tokens, registers filters, and updates CSS vars format

* Exposes `--output-references` CLI argument for `build-tokens` command. Defaults to `true`. Ensures brand package output with the CLI includes references in build output out-of-the-box.

* Registers filter(s) `isThemeVariant.{'light'}`, handling future theme variants when implemented (e.g., `isThemeVariant.dark`).
* Migrates `createCustomCSSVariables` to use `formattedVariables` to rely on out-of-the-box CSS variable formatting. The formatter still supports token-specific overrides of `outputReferences`. If a token has modifications via `modify`, the modified base reference is not included in the output.
* Updates custom fileHeader implementation, including a relative path to design tokens documentation.
* Fixes bug with line-height tokens, switching their `$type` from `dimension` to `number` to resolve typography style regressions.
* Updates typography tokens related to font size, font weight, and line-height for more consistent naming structure when taking into account mobile.
* Updates `@mobile-type` SCSS mixin to support level-specific customization of mobile typography styles for display 1-4.
* Renames `"description"` field in tokens to `"$description""` per the DTCG format.

* Ensures the "Typography" foundations page properly previews the correct font size for regular "Body" text and includes the missing "HEADING LABEL" example.
* Updates to "Colors" page in docs site:
  * Displays token name instead of CSS variable in the color swatch previews (see screenshot below).
  * Include `accent-a` and `accent-b` alongside other color names, rather than manually rendering `Swatch` for the accents.
  * Modifies the grid styles for color swatch preview to be more responsive.
* Resolves `NaNpx` bug in `MeasuredItem` component on docs site, while computing the measurements to display for an element (e.g., font size). Instead, it renders an empty block while measurements are resolved.
* Updates `CodeBlock` styles on docs site to add its border and background color only to the `LivePreview`, not the entire `CodeBlock` example.
* Reduces whitespace on docs site homepage.
* Simplifies columns on docs site header, ensuring `SiteTitle` is horizontally aligned in the center.

adamstankiewicz commented on tokens/src/core/global/typography.json:
```diff
@@ -2,90 +2,284 @@
+        "lg": {
           "source": "$lead-font-size",
-          "description": "Lead text font size."
+          "$value": "{typography.font.size.base} * 1.25",
```
D'oh, I think in my commit I lost the `calc(...)` that was previously wrapping this token's `$value` (this was the only token I experimented with). When previewing the docs site locally, the `.lead` utility class isn't working properly anymore, noticed first on the home page. Adding `calc(...)` back in will resolve.

[context] This was removed while exploring the `sd-transforms` transform for `ts/resolveMath` to see if we could abstract away `calc`, but as discussed in another comment, there is not currently a great solution for this, so we need to keep `calc` in here, at least for now :)

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3222

adamstankiewicz commented on tokens/src/core/components/Code.json:
```diff
@@ -1,32 +1,33 @@
   "typography": {
     "code": {
-      "font-size": { "value": "87.5%", "type": "percentage", "source": "$code-font-size" },
+      "font-size": { "source": "$code-font-size", "$value": "87.5%", "$type": "percentage" },
```
[suggestion] I believe this `87.5%` is equivalent to the `{typography.font.size.sm}` if we wanted to treat this token an alias to a semantic token (i.e., if/when `{typography.font.size.sm}` changes, this token will also change).

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3222

PKulkoRaccoonGang replied:
A small difference exists between `87.5%` and `0.875rem`. The `87.5%` value is calculated relative to the font size of the parent element. For example, if the parent has a font size of `16px`, then `87.5%` will be equivalent to `14px`. However, if the parent font size is different, the resulting pixel value will change accordingly.

In contrast, `0.875rem` is always calculated relative to the root element’s (`<html>`) base font size, which is typically `16px` by default in browsers. Thus, `0.875rem` will consistently be equivalent to `14px` as long as the html font size is `16px`, regardless of the parent element’s font size.

Currently, when using the `0.875rem` value from the token `{typography.font.size.sm}`, the value is indeed calculated relative to the root font size of the html tag (`16px` = `1rem`). The previous percentage value of `87.5%` was calculated based on the parent font size of `1.125rem` (the font size of the body).

I suggest using `87.5%`, what do you think?

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up commit: [6676868 (#3186)](https://github.com/openedx/paragon/pull/3186/commits/6676868e6c46d70d91f0bdc42b7c2848e3b0aa7c)

adamstankiewicz commented on tokens/src/core/global/typography.json:
```diff
@@ -2,90 +2,284 @@
+          "$value": "{typography.font.size.base} * 1.25",
+          "$description": "Lead text font size."
+        },
+        "sm": {
+          "source": "$small-font-size",
+          "$value": ".875rem",
+          "$description": "Small font size."
+        },
+        "xs": {
+          "source": "$x-small-font-size",
+          "$value": ".75rem",
+          "$description": "X-small font size."
```
Comparing the computed typography values in the docs for this PR's local checkout vs. published docs site, it looks like we might actually want/need to keep the `%` scale these tokens originally had.

For example, `x-small` used to be 75% but we migrated it to `.75rem`, which seems to result in slightly different font sizes. Note the `.small` and `.x-small` examples specifically (left: local; right: prod):

<img width="1899" alt="image" src="https://github.com/user-attachments/assets/0aa2dd69-9662-4b22-a0f1-9222ac081f81">

[aside] Related, you might notice the line heights are also slightly off for some of them as well. This seems to be due to the switch to `DataTable`, where some of the cell content is inheriting `DataTable`'s line-height styles, i.e. `.pgn__data-table td` (left: local; right: prod):

<img width="1911" alt="image" src="https://github.com/user-attachments/assets/f01d3623-b099-4537-9d01-a447b030dcd0">

I suppose we could consider overriding `DataTable`'s default CSS styles for typography example tables by `unset`'ing the line-height related definition(s), or maybe cross-check whether the CSS utility classes are defining explicit `line-height` style properties vs. inheriting from its parent.

PKulkoRaccoonGang replied:
Yes, we will definitely need to do a full check of all tokens after the latest changes, as we can easily miss some small detail that can have a significant impact on both the components and the documentation site. Thanks! 👍

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up commit: [6676868 (#3186)](https://github.com/openedx/paragon/pull/3186/commits/6676868e6c46d70d91f0bdc42b7c2848e3b0aa7c)

adamstankiewicz commented on tokens/src/core/global/typography.json:
```diff
@@ -2,90 +2,284 @@
+            "$value": "5.625rem",
+            "$description": "Font size for heading of level 3."
+          },
+          "4": {
```
On a local checkout of the docs site from this branch, it looks like the migrate of typography example tables to use `DataTable` is inheriting its `overflow: hidden` on table cells, cutting off the text of "Display 4" on smaller screen sizes. Probably not a big deal, but figured I'd flag it as it could be perceived as a style bug potentially.

<img width="1906" alt="image" src="https://github.com/user-attachments/assets/ea89a028-7914-42d5-b6e5-c46f1b5fbced">

<img width="196" alt="image" src="https://github.com/user-attachments/assets/48811055-b59e-416b-a73a-f445cdb1971f">

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up commit: [`6676868` (#3186)](https://github.com/openedx/paragon/pull/3186/commits/6676868e6c46d70d91f0bdc42b7c2848e3b0aa7c)

adamstankiewicz commented on tokens/src/core/components/general/link.json:
```diff
@@ -2,26 +2,27 @@
-        "base": { "value": "none", "type": "textDecoration", "source": "$link-decoration" },
-        "hover": { "value": "underline", "type": "textDecoration", "source": "$link-hover-decoration" },
+        "$type": "textDecoration",
+        "base": { "source": "$link-decoration", "$value": "none" },
```
[inform, docs site bug] The "Links" section on the typography foundations page is incorrectly showing `text-decoration: underline;` on "standalone links" (left: local; right: prod):

<img width="1899" alt="image" src="https://github.com/user-attachments/assets/52d6c4ff-dd18-44bb-9f21-67e9c2bcae93">

It appears this is due to the migration of the example tables to `DataTable`, where the `TextCell` is wrapping the cell content with a `p`, which according to the table description, "For links inside a `p`...", the standalone links are actually getting treated as inline links even without the explicit `.inline-link` class.

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up commit: [`6676868` (#3186)](https://github.com/openedx/paragon/pull/3186/commits/6676868e6c46d70d91f0bdc42b7c2848e3b0aa7c)

PKulkoRaccoonGang committed:
feat: added expanded tokens

refactor: corrected expanded tokens

refactor: code refactoring

refactor: created an abstracted CSS variables

refactor: added missing expanded tokens

refactor: added missing expanded tokens

refactor: small refactoring after review

refactor: corrected imports for CSS files

feat: expanded --pgn-transition-carousel-base token, some refactoring

fix: corrected --pgn-transition-base-timing-function value

refactor: some refactoring and updates

refactor: corrected progress-bar and typography tokens

adamstankiewicz committed:
feat: added sd-transforms; refs in source-tokens-only output; fixed refs warnings

PKulkoRaccoonGang committed:
fix: fixed typography, display and links sections

PKulkoRaccoonGang commented on www/src/components/typography-page/Links.tsx:
```diff
@@ -34,6 +31,9 @@ export default function Links() {
       <h2 className="mb-2">Links</h2>
+      <Alert variant="info" icon={Info}>
+        <a href="/#">Standalone Link</a> - the default style for <code>a</code> tag that appear outside of <code>p</code> tag.
+      </Alert>
```
<img width="961" alt="image" src="https://github.com/user-attachments/assets/ba409567-f1a8-4a58-bbe8-be1decdbe812">

PKulkoRaccoonGang commented:
@brian-smith-tcril @adamstankiewicz In [this commit](https://github.com/openedx/paragon/pull/3186/commits/da00992533e1ac51b1e9bd53b43e8c50239e8a35#diff-1b8d91c59681705b700c37ce0d260f55ab525bf6174bb2686ecc37873e1897e0L1) we removed the brand-openedx theme. Do we want to remove the sidebar toggle button and any mention of the brand in Paragon?

<img width="235" alt="image" src="https://github.com/user-attachments/assets/41fe497f-58ea-4569-ae3e-c11b08756563">

brian-smith-tcril commented on tokens/src/themes/light/components/Carousel.json:
```diff
@@ -1,43 +1,44 @@
             "outputReferences": false,
-            "modify": [{ "type": "str-replace", "toReplace": "#", "replaceWith": "%23" }]
+            "modify": [{ "type": "str-replace", "toReplace": "#", "replaceWith": "%23" }],
+            "$value": "url(\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='{color.carousel.control.base}' viewBox='0 0 8 8'%3e%3cpath d='M5.25 0l-4 4 4 4 1.5-1.5-2.5-2.5 2.5-2.5-1.5-1.5z'/%3e%3c/svg%3e\")"
```
I'm curious about the hardcoded `%3c` for `<` and `%3e` for `>` here. I'm not 100% sure how `str-replace` in `modify` works, could that be used for those similarly to how it's being used for `#`?

PKulkoRaccoonGang replied:
This is a [modifier](https://github.com/openedx/paragon/blob/alpha/tokens/style-dictionary.js#L112) for replacing one value in a string with another. After playing with this modifier a bit, I realized that it doesn't support multiple replacement parameters, for example, you can't use something like:

```
            "modify": [
              { "type": "str-replace", "toReplace": "#", "replaceWith": "%23" },
              { "type": "str-replace", "toReplace": ">", "replaceWith": "%3E" },
              { "type": "str-replace", "toReplace": "<", "replaceWith": "%3C" }
            ],
            "$value": "url(\"data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' fill='{color.carousel.control.base}' viewBox='0 0 8 8'><path d='M2.75 0l-1.5 1.5 2.5 2.5-2.5 2.5 1.5 1.5 4-4-4-4z'/></svg>\")"
```

I'm not sure if we need to handle any cases right now other than the `#` symbol (we're currently replacing only that symbol), but I think the option of multiple parameters for replacement would be useful in the future 💯

brian-smith-tcril replied:
That's unfortunate. It would be really nice to support multiple `str-replace`s. Not a dealbreaker though.

brian-smith-tcril reviewed (commented):
Overall this is looking really great! I left a few comments with questions, but it's awesome to see v4 so close to ready!

adamstankiewicz commented on bin/paragon-scripts.js:
```diff
@@ -85,11 +85,21 @@ const COMMANDS = {
       },
+      {
+        name: '--output-references',
+        description: 'Include references in the build output.',
```
@brian-smith-tcril Would it be helpful to include "Include **token** references in the build output"? This is getting at whether a token who's `$value` references another token should result in the CSS file output as a CSS variable or a hardcoded value. 

`--output-references` == include CSS variables for tokens with aliases to other tokens.

adamstankiewicz replied:
Related, if that distinction of **token reference** is helpful, could rename the argument flag altogether to `--include-token-references`.

brian-smith-tcril replied:
I think I like "Include CSS variables for tokens with aliases to other tokens." as the description. I'm fine with keeping the argument as `--output-references`, but `--include-token-references` or `--output-token-references` work too.

adamstankiewicz replied:
My concern about mentioning "CSS variables" specifically is that the long term version is that we're not just outputting CSS variables, but in the future also formatted tokens for iOS/Android, etc. where "CSS variables" doesn't make much sense.

Similarly, if a consumer were to extend the default `style-dictionary` config to export SCSS variables from the tokens (beyond the default CSS variables), the `--output-tokens-references` is referencing SCSS variables, not CSS variables.

brian-smith-tcril replied:
Good point, would just, "Include variables for tokens with aliases to other tokens." work?

adamstankiewicz replied:
I'd be OK with that, though the slight tradeoff is that we'd be slightly moving away from style-dictionary's official language around "references".

> Include variables for tokens with references to other tokens."

🤔

brian-smith-tcril replied:
I was actually thinking that too. My main problem with "Include references in the build output." is that it doesn't explain what references mean or what said references may be referring to. I'd be happy with "Include references for tokens with aliases to other tokens in the build output."

adamstankiewicz replied:
Sold, let's roll with that :) 

> Include references for tokens with aliases to other tokens in the build output.

brian-smith-tcril replied:
So now the only question is the argument name itself, any strong feelings between
* `--output-references`
* `--include-token-references`
* `--output-token-references`
or maybe even
* `--include-references`

adamstankiewicz replied:
I think `--output` makes sense since it's controlling the output build files, whereas `--include` _could_ suggest including references pre-output.

I tend to prefer `--output-token-references`  the most, but don't feel too strongly.

brian-smith-tcril replied:
Sounds good to me, let's go with `--output-token-references`!

PKulkoRaccoonGang replied:
Nice! This will definitely make more sense now 💯

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up commit: [`036f28c` (#3186)](https://github.com/openedx/paragon/pull/3186/commits/036f28cf6023eb750ee198dc010186f18101b533)

adamstankiewicz commented on lib/build-tokens.js:
```diff
@@ -41,33 +59,35 @@ async function buildTokensCommand(commandArgs) {
     platforms: {
       css: {
         prefix: 'pgn',
-        transformGroup: 'css',
```
Confirmed, this conversation can be resolved 👍

adamstankiewicz commented on lib/build-tokens.js:
```diff
@@ -38,36 +57,42 @@ async function buildTokensCommand(commandArgs) {
         // NOTE: buildPath must end with a slash
         buildPath: buildDir.slice(-1) === '/' ? buildDir : `${buildDir}/`,
+        options: {
+          fileHeader: 'customFileHeader',
```
The following is the `customFileHeader` config:

```
/**
 * Registers a custom fileHeader.
 */
StyleDictionary.registerFileHeader({
  name: 'customFileHeader',
  fileHeader: (defaultMessage) => [
    `${defaultMessage} while transforming design tokens.`,
    'See <root>/tokens/README.md for more details.',
  ],
});
```

IIRC, the current file header is where we want to be for now (i.e., not including timestamps, including a docs reference to learn more).

brian-smith-tcril replied:
Awesome! Thanks for remembering/confirming!

brian-smith-tcril commented on styles/css/core/custom-media-breakpoints.css:
```diff
@@ -1,10 +1,9 @@
- * IMPORTANT: This file is the result of assembling design tokens
- * Do not edit directly
- * Generated on Fri, 22 Sep 2023 08:20:17 GMT
+ * Do not edit directly, this file was auto-generated. while transforming design tokens.
+ * See <root>/tokens/README.md for more details.
```
Resolved as per https://github.com/openedx/paragon/pull/3186#discussion_r1823114970

adamstankiewicz replied:
Yep, this is the intended custom file header we landed on after a few iterations (i.e., no timestamp, include reference to README docs for more details).

adamstankiewicz commented on styles/scss/core/_utilities.scss:
```diff
@@ -21,13 +21,17 @@
 
 .x-small {
-  font-size: $x-small-font-size;
+  font-size: $x-small-font-size !important;
```
CSS utility classes in the design system _should_ have `!important`; check other existing CSS utility classes for reference, e.g.:

<img width="220" alt="image" src="https://github.com/user-attachments/assets/c282c629-f0f3-4ce5-90c2-4fcd3cf04916">

brian-smith-tcril replied:
> CSS utility classes in the design system should have `!important`

That's what I thought - I guess I'm just not sure why that wasn't the case *before*, and if there might be something that currently relies on them unconventionally *not* having `!important`. I'm just worried about a possible "if it's a bug people rely on it's a feature" scenario.

PKulkoRaccoonGang replied:
I searched on GitHub for the use of certain utilities in the OpenEdX repositories and did not find any user overrides that could be impacted by the use of `!important`. My vote is to keep `!important`.

brian-smith-tcril replied:
Thanks for going above and beyond and looking into that @PKulkoRaccoonGang!

adamstankiewicz replied:
Appreciate the sanity check, @PKulkoRaccoonGang!

adamstankiewicz commented on lib/build-tokens.js:
```diff
@@ -85,49 +110,60 @@ async function buildTokensCommand(commandArgs) {
             format: 'css/custom-variables',
             destination: `themes/${themeVariant}/variables.css`,
-            filter: hasSourceTokensOnly ? 'isSource' : undefined,
+            filter: hasSourceTokensOnly
+              ? `isSource.${themeVariant}`
+              : `isThemeVariant.${themeVariant}`,
```
This specific `filter` pertains to theme variants, regardless of `hasSourceTokensOnly`.

For example, `isSource.${themeVariant}` is filtering for source tokens of a specific theme variant; similarly, `isThemeVariant.${themeVariant}` is filtering for all tokens (source or otherwise) for a specific theme variant.

`source` tokens are the ones from custom brand packages, i.e. the tokens to extend/modify the default tokens defined within the Paragon repo (passed via style-dictionary's `include`).

When `hasSourceTokensOnly`, it's telling style-dictionary to only transform/output tokens from the `source` tokens (aka only the custom brand package tokens, not the full set of tokens in upstream Paragon).

brian-smith-tcril replied:
> When `hasSourceTokensOnly`, it's telling style-dictionary to only transform/output tokens from the source tokens (aka only the custom brand package tokens, not the full set of tokens in upstream Paragon).

I guess the word that feels a little odd to me is "has" - maybe `transformSourceTokensOnly` or `outputSourceTokensOnly`?

adamstankiewicz replied:
Fair enough. I like the transform prefix suggestion instead 🙂

brian-smith-tcril replied:
I like `transformSourceTokensOnly` the best so far too! Let's roll with that!

brian-smith-tcril replied:
https://github.com/openedx/paragon/pull/3186/commits/b4d560bee10f28a9ea2c9d26275b647334a00111

adamstankiewicz commented on tokens/src/core/components/Avatar.json:
```diff
@@ -1,16 +1,16 @@
-      "xl": { "value": "6rem", "type": "dimension", "source": "$avatar-size-xl" },
-      "xxl": { "value": "11.5rem", "type": "dimension", "source": "$avatar-size-xxl" },
-      "huge": { "value": "18.75rem", "type": "dimension", "source": "$avatar-size-huge" },
+      "base": { "source": "$avatar-size", "$value": "3rem", "$type": "dimension" },
+      "xs": { "source": "$avatar-size-xs", "$value": "1.5rem", "$type": "dimension" },
+      "sm": { "source": "$avatar-size-sm", "$value": "2.25rem", "$type": "dimension" },
+      "lg": { "source": "$avatar-size-lg", "$value": "4rem", "$type": "dimension" },
+      "xl": { "source": "$avatar-size-xl", "$value": "6rem", "$type": "dimension" },
+      "xxl": { "source": "$avatar-size-xxl", "$value": "11.5rem", "$type": "dimension" },
+      "huge": { "source": "$avatar-size-huge", "$value": "18.75rem", "$type": "dimension" },
```
super nit: might recommend keeping `source` the last property defined, as the other properties like `$value`, and `$type` are more critical.

PKulkoRaccoonGang replied:
Agreed, that's a good idea 👍 
I would also leave some documentation somewhere about what rules to follow when creating/editing design tokens. Since it's not entirely obvious what sequence should be present in design tokens. I think we should stop at something like this:
1. `$value`
2. `description`
3. `modify`
4. `source`

PKulkoRaccoonGang replied:
[inform]: We have a draft PR providing additions of a linter for JSON files and a sorter script
https://github.com/openedx/paragon/pull/3285

adamstankiewicz commented on tokens/src/core/components/CloseButton.json:
```diff
@@ -2,12 +2,8 @@
-        "weight": {
-          "value": "{typography.font.weight.bold}", "type": "fontWeight", "source": "$close-font-weight"
-        }
+        "size": { "source": "$close-font-size", "$value": "calc({typography.font.size.base} * 1.5)", "$type": "dimension" },
```
See https://github.com/openedx/paragon/pull/3186/files#r1739426413 for details on where we landed regarding `calc.`

tl;dr; there's no great alternative right now to avoid `calc` in style-dictionary tokens.

PKulkoRaccoonGang replied:
How about creating an issue to track a specific CSS `calc()`?
This way, we won’t lose track of it in the future.

adamstankiewicz replied:
Thanks @brian-smith-tcril!

PKulkoRaccoonGang commented on lib/build-tokens.js:
```diff
@@ -38,36 +57,42 @@ async function buildTokensCommand(commandArgs) {
       },
     },
+    log: {
+      verbosity: verbose ? 'verbose' : 'default',
```
There is no significant difference in the case of displaying errors. I would leave default as the default value (with the option to show more verbose logs)

Default example
<img width="1061" alt="Screenshot 2024-11-01 at 11 20 43" src="https://github.com/user-attachments/assets/9928c719-fdbd-4089-93d1-ee96f0d6fe2b">

Verbose example
<img width="1042" alt="image" src="https://github.com/user-attachments/assets/2c723b60-ae76-4191-9951-0e2096c2424a">

brian-smith-tcril replied:
Sounds good to me. If someone comes up with a use case for needing to silence logs (none come to mind), or `style-dictionary` is updated to have more verbosity options, we can look into changing this in the future.

PKulkoRaccoonGang committed:
chore: renamed --output-token-references

PKulkoRaccoonGang commented on src/Card/CardDeck.jsx:
```diff
@@ -1,7 +1,6 @@
 import React, { Children, useMemo } from 'react';
 import classNames from 'classnames';
 import PropTypes from 'prop-types';
-import BaseCardDeck from 'react-bootstrap/CardDeck';
```
Do you mean adding the removal of the missed deprecated Card components to an [existing commit](https://github.com/openedx/paragon/pull/3158/commits/5a23e5740dd35fa15b6166f34079620e231e9fee) that was made earlier or moving this change to a separate PR/commit?

brian-smith-tcril replied:
At first I thought there were more of these in this PR. Since there aren't a ton I'm fine with keeping this in this PR.

PKulkoRaccoonGang committed:
refactor: renamed hasSourceTokensOnly variable

adamstankiewicz commented on tokens/src/core/global/elevation.json:
```diff
@@ -1,19 +1,61 @@
 {
   "elevation": {
+    "$type": "ratio",
```
[sanity check] Does `ratio` make sense as the type (see [DTCG format spec](https://tr.designtokens.org/format/#additional-types)'s description of the `ratio` type)? Perhaps we want this to be `"$type": "number"` since it's really just a number ([docs](https://tr.designtokens.org/format/#number))?

PKulkoRaccoonGang replied:
[inform] Implemented in follow-up PR: https://github.com/openedx/paragon/pull/3286

adamstankiewicz commented on tokens/src/core/alias/size.json:
```diff
@@ -1,15 +1,16 @@
+    "$type": "dimension",
     "border": {
-      "width": { "value": "1px", "type": "dimension", "source": "$border-width", "description": "Default border width." },
+      "width": { "source": "$border-width", "$value": "1px", "$description": "Default border width." },
```
[curious] Cross-checking the `dimension` type with the DTCG spec again, I noticed that it seems you can also define the dimension value and its unit separately, e.g.:

```json
{
  "size": {
    "$type": "dimension",
    "border": {
      "width": {
        "$value": {
          "value": 1,
          "unit": "px"
        }
      }
     }
  }
}
```

Not sure if this is something `style-dictionary` v4 supports out-of-the-box. Do we think it's worth exploring defining dimension token values and units separately vs. combined as we have it today (i.e., `"$value": "1px"`).

brian-smith-tcril replied:
Since this is what is documented in the DTCG spec I'd think we'd want to move forward with this change. Let's verify this works with style-dictionary v4 and figure out next steps from there.

adamstankiewicz replied:
For posterity, we tried to update the token definition to adopt the DTCG format spec for `$dimension` tokens, but it seems to result in unexpected output, e.g.:

```css
--pgn-size-border-width-value: 1; /* Default border width. */
--pgn-size-border-width-unit: 0px; /* Default border width. */
```

When we'd really want something like:

```css
--pgn-size-border-width-value: 1px; /* Default border width. */
```

I've opened an issue upstream in the `style-dictionary` repository to try to drive clarity in the expected output for `$dimension` tokens using the DTCG spec / report a bug, if applicable: https://github.com/amzn/style-dictionary/issues/1398

Given this, for now, we will keep the Paragon `$dimension` tokens using the combined `$value` vs. following the DTCG spec.

PKulkoRaccoonGang committed:
feat: Removed default value for isOverflowVisible in ModalDialog component

PKulkoRaccoonGang committed:
refactor: corrected ratio tokens type and renamed edx theme

brian-smith-tcril reviewed (approved):
I believe all my previous comments have been addressed. This looks great!

Super excited to see all of this come together. Amazing work @PKulkoRaccoonGang!

adamstankiewicz reviewed (approved):
Looks great! Appreciate all your iterations on this PR, @PKulkoRaccoonGang! Will be great to see this land.
